### PR TITLE
Make all plugin events registered to make documenting them easier

### PR DESCRIPTION
**Pull request by jules2689**
# What?

Every event in #hook and #add_hook will check if the event is registered in
Bundler::Plugin::Events. This allows for easy tracking of what's calling events,
and allow documentation to easily point to a single location.

It also makes testing easier as events are predicatable and accessible via constants

Thanks so much for the contribution!
To make reviewing this PR a bit easier, please fill out answers to the following questions.

---

### What was the end-user problem that led to this PR?

There wasn't really a problem, but this makes adding events easier

### What was your diagnosis of the problem?

Events are not tracked or documented well, this makes them documented well

### What is your fix for the problem, implemented in this PR?

Add an Events registration that hooks test for

### Why did you choose this fix out of the possible options?

Constants make things easily accessible via code, the hash makes it easy to check (with `O(1)` access!) for an event based on value as well.

### CODE REVIEW

File: lib/bundler/plugin/events.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module Bundler
+  module Plugin
+    module Events
+      def self.define(const, event)
```
[the remark this answers is not in the document]
Comment (jules2689) [reply lib/bundler/plugin/events.rb]: done

**Comment by jules2689** [comment]
cc @segiddins 
I opted to use a `@!parse` block to document the constants instead of a macro as it seemed to work better and made constant docs rather than a method doc.

**Comment by segiddins** [comment]
I’m pretty sure a macro could’ve defined the constant, so we could reduce duplication? And then the parse block could go inside the macro? Not super important

**Comment by jules2689** [comment]
> I’m pretty sure a macro could’ve defined the constant, so we could reduce duplication? And then the parse block could go inside the macro? Not super important

I couldn't find a way to accomplish this :( I'm not super familiar with macros

File: spec/bundler/plugin_spec.rb [hunk 2]
```diff
@@ -251,14 +251,29 @@
...
+
   describe "#hook" do
     before do
       path = lib_path("foo-plugin")
       build_lib "foo-plugin", :path => path do |s|
         s.write "plugins.rb", code
       end
 
-      allow(index).to receive(:hook_plugins).with(event).
+      Bundler::Plugin::Events.send(:define, :EVENT_1, "event-1")
```
Comment (segiddins) [inline spec/bundler/plugin_spec.rb]: I think in a spec_helper we may want to reset the set of `events` so there's no possibility for test ordering dependence

**Comment by colby-swandale** [comment]
@bundlerbot r+

**Comment by colby-swandale** [comment]
woo!! :tada:

**Comment by colby-swandale** [comment]
@bundlerbot retry

**Pull request merged**